File: sim/spore.ts
import { Entity } from './entity'
import { SporeMode, Root, EntityState } from './types'
import { Vector } from './vector'
import { Land } from './land'

export class Spore extends Entity {
  water: number
  nutrients: number
  energy: number

  readonly requiredEnergy = 250
  readonly rootEnergy = 100

  waterSuction: number
  nutrientSuction: number

  mode: SporeMode
  lifeTime: number

  growthDir: Vector
  segLength: number
  rootPropDist: number

  roots: Root[]

  landId: number | null
  shouldRoot: number
  attemptedRootFailed: boolean

  constructor(x: number, y: number) {
    super('spore')
    this.x = x; this.y = y
    this.radius = 5
    this.mass = 5
    this.maxSpeed = 30
    this.charge = 1

    this.water = 0
    this.nutrients = 0
    this.energy = 0
    this.waterSuction = 5
    this.nutrientSuction = 5

    this.mode = 'dormant'
    this.lifeTime = 0

    this.growthDir = new Vector(1, 1, 0)
    this.growthDir.makeUnit()
    this.segLength = 4
    this.rootPropDist = 10

    this.roots = []
    this.landId = null
    this.shouldRoot = 0
    this.attemptedRootFailed = false
  }

  simulate(world: import('./world').World): void {
    if (this.lifeTime > 1000) {
      this.alive = false
      return
    }

    if (this.mode === 'dormant') {
      const land = world.getLand(this.x, this.y)
      if (land && land.fertility > 0.7) {
        this.germinate(land)
      }
    } else {
      this.forceStop()
      this.absorbNutrients(world)
      this.osmosis(world)
      this.produceEnergy()

      if (this.energy >= this.requiredEnergy) {
        this.attemptedRootFailed = false
        world.spawnPlant(this.x, this.y, this.growthDir.x, this.growthDir.y, this.segLength)
        this.alive = false
        return
      }

      const land = world.getLand(this.x, this.y)
      if (land && land.fertility < 0) {
        const prop = this.energy / this.requiredEnergy
        if (prop < 0.5) this.shouldRoot += 0.05
        if (this.shouldRoot > 0.75) {
          if (!this.growRoot(world)) this.attemptedRootFailed = true
        }
      }
      if (this.attemptedRootFailed) {
        if (this.growRoot(world)) {
          this.attemptedRootFailed = false
          this.shouldRoot = 0
        }
      }
    }
  }

  private germinate(land: Land): void {
    if (this.mode !== 'dormant') return
    this.mode = 'growing'
    this.landId = land.id
    this.attachToPoint(this.x, this.y, 0, 1000, 0)

    const spd = this.speed
    if (spd > 0) {
      this.growthDir = this.getVel().normalize()
    }
    this.forceStop()
  }

  private osmosis(world: import('./world').World): void {
    const land = world.getLandById(this.landId)
    if (land) {
      const before = land.moisture
      land.changeMoisture(-this.waterSuction)
      this.water += before - land.moisture
    }
    for (const root of this.roots) {
      const rLand = world.getLandById(root.landId)
      if (rLand) {
        const before = rLand.moisture
        rLand.changeMoisture(-root.waterSuction)
        this.water += before - rLand.moisture
      }
    }
  }

  private absorbNutrients(world: import('./world').World): void {
    const land = world.getLandById(this.landId)
    if (land) {
      const before = land.nutrients
      land.changeNutrients(-this.nutrientSuction)
      this.nutrients += before - land.nutrients
    }
    for (const root of this.roots) {
      const rLand = world.getLandById(root.landId)
      if (rLand) {
        const before = rLand.nutrients
        rLand.changeNutrients(-root.nutrientSuction)
        this.nutrients += before - rLand.nutrients
      }
    }
  }

  private produceEnergy(): void {
    // 10 water + 15 nutrients -> 5 energy
    if (this.water >= 10 && this.nutrients >= 15) {
      this.water -= 10
      this.nutrients -= 15
      this.energy += 5
    }
  }

  private growRoot(world: import('./world').World): boolean {
    if (this.energy < this.rootEnergy) return false

    const dx = this.growthDir.scale(this.rootPropDist)
    const nx = this.x + dx.x
    const ny = this.y + dx.y

    if (nx < 0 || nx >= world.width || ny < 0 || ny >= world.height) {
      this.growthDir.rotateInPlace2D(5 * Math.PI / 180)
      return false
    }

    const land = world.getLand(nx, ny)
    if (!land) {
      this.growthDir.rotateInPlace2D(5 * Math.PI / 180)
      return false
    }

    this.energy -= this.rootEnergy
    this.roots.push({ x: nx, y: ny, landId: land.id, waterSuction: this.waterSuction, nutrientSuction: this.nutrientSuction })
    this.growthDir.rotateInPlace2D(5 * Math.PI / 180)
    return true
  }

  override toJSON(): EntityState {
    return {
      ...super.toJSON(),
      type: 'spore',
      energy: this.energy,
      mode: this.mode,
    }
  }
}

File: game_cpp/src/spore.cpp
#include "spore.h"

spore::spore(POINT loc, world* WORLD)
{
	water = 0;
	nutrients = 0;
	energy = 0;
	required_energy = 250;
	root_energy = 100;
	water_suction = 5;
	nutrient_suction = 5;
	should_root = 0;
	attempted_root_failed = false;
	seg_length = 4;
	root_prop_dist = 10;

	//growth direction is becomes direction of velocity when spore germinates.
	growth_dir = Vector(1, 1, 0);


	pWorld = WORLD;
	pLand = nullptr;
	life_time = 0;


	alive = true;
	view_loc = loc;
	part_loc = winPoint_to_myPoint(loc);
	max_speed = 30;
	radius = 5;
	type = "spore";
}
void spore::simulate()
{
	if (life_time > 1000)
	{
		//spore has reached the end of its lifespan.
		alive = false;
	}
	else
	{
		if (mode == spore_mode::DORMANT)
		{
			//Search for suitable conditions, take root when such are found.
			land* curLand = pWorld->getLand(view_loc);
			if (curLand)
			{
				if (curLand->fertility > .7)
				{
					//Suitable conditions.
					germinate(curLand);
				}
			}
		}
		else if (mode == spore_mode::GROWING)
		{
			//Absorb nutrients/water, produce energy, spawn plant when possible.
			//Consider this a simulation of an attachment to ground.
			force_stop();	//prevent wind from pushing spore.

			absorb_nutrients();
			osmosis();
			produce_energy();

			//Try to spawn plant.
			int diff = energy - required_energy;
			if (diff >= 0)
			{
				//energy sufficient. -> forget about rooting and grow plant.
				attempted_root_failed = false;
				pWorld->spawn_plant(view_loc, growth_dir, seg_length);
				alive = false;	//set this spore to dead so it will be removed by the world.
			}

			//Setting it to less than zero for now so that it won't do it because the mechanics need work.
			//Roots definitely help it to gather energy faster, however, when it keeps making roots instead of just saving energy, it'll never spawn a plant.
			if (pLand->fertility < 0)
			{
				float prop = energy / required_energy;
				if (prop < .5)
				{
					should_root += .05;
				}
				if (should_root > .75)
				{
					//Fertility of land has been very low while spore has been far from the requirement 17 times.
					//Growing a new root calculated as a good idea.
					if (!grow_root())
					{
						attempted_root_failed = true;
					}
				}
			}
			//The following represents the intention of the spore to generate a new root once it has gathered enough energy/found room.
			if (attempted_root_failed)
			{
				if (grow_root())
				{
					attempted_root_failed = false;
					should_root = 0;
				}
			}





		}
	}
}
void spore::germinate(land* curLand)
{
	if (mode == spore_mode::DORMANT)
	{
		mode = spore_mode::GROWING;
		pLand = curLand;

		//Spore also attaches itself to the land via the particle point.
		attach_to_point(part_loc, 1000, 0);

		if (vel.get_size() > 0)
		{
			growth_dir = vel;	//growth_direction equals current velocity.
			growth_dir.make_unit();
			//If velocity happens to be zero, then we'll leave the growth dir at the default.
			//Perhaps in future, it will be in the direction of the most fertile adjacent land plot.
		}
		force_stop();

	}
}
bool spore::grow_root()
{
	//Costs energy, will delay plant spawning. Spore should calculate when to use this.
	if (energy >= root_energy)
	{	
		//Sufficient energy.
		Vector dx = growth_dir;
		dx = dx.MultiplyVectorByScalar(root_prop_dist);
		POINT nLoc = myPoint_to_winPoint(part_loc.AddVectorToPoint(&dx));

		bool enough_room = PtInRect(&pWorld->brc, nLoc);

		if (enough_room)
		{
			//Sufficient room.
			//Then it's possible to form root.	
			energy -= root_energy;
			land* nLand = pWorld->getLand(nLoc);
			root nRoot(nLoc, nLand, water_suction, nutrient_suction);
			roots.push_back(nRoot);

			//Rotate growth_direction (so that next roots won't go to same place... one way or another, plants do (smth. like) this, with spiralling as well -- though I'll ommit that detail for now).
			myMatrix mat;
			mat.init_RotXY_3x3(5);
			mat.Rotate_XY_3x3(&growth_dir, &growth_dir);

			return true;


		}

		//Rotate growth_direction (so that next roots won't go to same place... one way or another, plants do (smth. like) this, with spiralling as well -- though I'll ommit that detail for now).
		myMatrix mat;
		mat.init_RotXY_3x3(5);
		mat.Rotate_XY_3x3(&growth_dir, &growth_dir);
		return false;
	}
	return false;
}

void spore::absorb_nutrients()
{
	if (pLand)
	{
		float nutrient_initial = pLand->nutrient_level;
		pLand->change_nutrients(-1 * nutrient_suction);
		float nutrient_final = pLand->nutrient_level;
		float delta_nutrient = nutrient_final - nutrient_initial;
		float absorbed_nutrient = -1 * delta_nutrient;
		nutrients += absorbed_nutrient;

	}
	else
	{
		//Error: no land.
	}

	//absorb with roots
	for (int i = 0; i < roots.size(); i++)
	{
		if (roots[i].pLand)
		{
			float nutrient_initial = roots[i].pLand->nutrient_level;
			roots[i].pLand->change_nutrients(-1 * roots[i].nutrient_suction);
			float nutrient_final = roots[i].pLand->nutrient_level;
			float delta_nutrient = nutrient_final - nutrient_initial;
			float absorbed_nutrient = -1 * delta_nutrient;
			nutrients += absorbed_nutrient;
		}
		else
		{
			//This root's land pointer is null.
		}
	}
}
void spore::osmosis()
{
	if (pLand)
	{
		float moisture_initial = pLand->moisture_level;
		pLand->change_moisture(-1 * water_suction);
		float moisture_final = pLand->moisture_level;
		float delta_water = moisture_final - moisture_initial;
		float absorbed_water = -1 * delta_water;
		water += absorbed_water;
	}
	else
	{
		//Error: no land.
	}
	//absorb with roots.
	for (int i = 0; i < roots.size(); i++)
	{
		if (roots[i].pLand)
		{
			float moisture_initial = roots[i].pLand->moisture_level;
			roots[i].pLand->change_moisture(-1 * roots[i].water_suction);
			float moisture_final = roots[i].pLand->moisture_level;
			float delta_water = moisture_final - moisture_initial;
			float absorbed_water = -1 * delta_water;
			water += absorbed_water;
		}
		else
		{
			//This root's land pointer is null.
		}
	}
}
void spore::produce_energy()
{
	//10wat + 15nut --> 5ener

	if (water - 10 >= 0)
	{
		if (nutrients - 15 >= 0)
		{
			water -= 10;
			nutrients -= 15;
			energy += 5;
		}
	}

}


void spore::draw(HDC hdc, HWND hwnd)
{
	HPEN whitePen = CreatePen(PS_SOLID, 2, COLORREF(0xFFFF00));
	HBRUSH whiteBrush = CreateSolidBrush(COLORREF(0xFFFF00));
	RECT brc;
	brc.left = view_loc.x - radius;
	brc.right = view_loc.x + radius;
	brc.top = view_loc.y - radius;
	brc.bottom = view_loc.y + radius;
	SelectObject(hdc, whitePen);
	SelectObject(hdc, whiteBrush);
	Ellipse(hdc, brc.left, brc.top, brc.right, brc.bottom);
	DeleteObject(whitePen);
	DeleteObject(whiteBrush);

	for (int i = 0; i < roots.size(); i++)
	{
		HPEN brownPen = CreatePen(PS_SOLID, radius, COLORREF(0x082945));
		SelectObject(hdc, brownPen);

		MoveToEx(hdc, view_loc.x, view_loc.y, NULL);
		LineTo(hdc, roots[i].location.x, roots[i].location.y);
		DeleteObject(brownPen);
	}

}